### ISSUE TITLE
[dev.icinga.com #12688] Having tmp and perfdata directories on different fs not working

### ISSUE THREAD

**Issue by icinga-migration** [Icinga/icinga2#4660]
This issue has been migrated from Redmine: https://dev.icinga.com/issues/12688

**Created by tontonitch on 2016-09-10 10:17:01 +00:00**

Assignee: _(none)_
Status: _New_ 
Target Version: _(none)_
Last Update: _2016-09-18 13:32:31 +00:00 (in Redmine)_

    Icinga Version: 2.5.4
    Backport?: Not yet backported
    Include in Changelog: 1

---

Hi,

Facing currently a load and I/O issue with perfdata generation/processing, I'm trying to optimize the perfdata processing.
For that, I'm trying to use tmpfs for some directories.
The icinga2 related directories are:

    /var/spool/icinga2/tmp
    /var/spool/icinga2/perfdata

1. First I've tried to create a single tmpfs mounted to /var/spool/icinga2
Unfortunately, the tmp and perfdata directories are not created at icinga2 startup and should be created manually.

2. Then, I created 2 separate tmpfs+mountpoints

    tmpfs                                       50M  108K   50M   1% /var/spool/icinga2/tmp
    tmpfs                                       50M     0   50M   0% /var/spool/icinga2/perfdata

Icinga2 is happy with that, until trying to move the perfdata file from tmp to perfdata dir.

    [2016-09-10 12:01:32 +0200] critical/ThreadPool: Exception thrown in event handler:
    Error: Function call 'rename' for file '/var/spool/icinga2/tmp/host-perfdata' failed with error code 18, 'Invalid cross-device link'

Indeed, it uses the rename function, which doesn't works across fs.
See http://www.linuxquestions.org/questions/programming-9/rename-problem-429319/

So, to be more flexible, it would be nice to be able to store tmp and perfdata related files to different fs.
One alternative, linked to 1. would be that icinga2 creates the tmp and perfdata directories at startup, so we could use a single tmpfs+mountpoint.

Best regards,
Yannick

**Comment by Thomas-Gelf** [comment]
I know, I repeat myself: please don't. Instead of bringing better performance this would perform much, much worse. @gunnarbeutner already explained that this would force us to do I/O, so no chance to get faster this way. And getting faster was the initial motivation for @Tontonitch. Just, this goal cannot be achieved by replacing a single atomic operation with userspace I/O.

Even worse, this has the potential to break many things by introducing race conditions. Methods like `PerfdataWriter::RotateFile` have been implemented to protect against such. Otherwise writing temporary files first (before moving them in an atomic way) would be pointless. The code in the linked pull request is extremely dangerous for an application like Icinga, where a lot of concurrency is going on.

Long story short, please reject this feature request. We are not doing our users a favor by allowing them to do hurt themselves. The desire to duplicate perfdata I/O makes a "perfect" example here.

The pull-request itself helps to abstract an often-used method, this absolutely makes sense to me. Still, `Utility::MoveFile` should not be added without a flag allowing one to enforce atomicity. Instead of using a flag, it could also be complemented by an atomic `Utility::RenameFile`. That one should be the default for most calls in our code. And it should of course log/throw errors where appropriate, as proposed by @Al2Klimov.

Cheers,
Thomas

**Comment by gunnarbeutner** [comment]
As explained by @Thomas-Gelf moving temporary files across device boundaries is not something se should encourage users to do.
